Cypress test. OpenCV basic actions. Histogram Equalization.

dvkruchinin opened the pull request:
<!---
Copyright (C) 2020-2021 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It will help avoiding duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
Cypress tests for features.

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

bsekachev commented:
@dvkruchinin Please, merge develop branch to fix remark issues.

bsekachev commented:
@klakhov 
Please, look at the test. Try to run it via cypress, comment issues if you see them, let me know if you have difficulties.

klakhov reviewed (dismissed):
I've run it via cypress, everything seems fine.

bsekachev commented on cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/opencv-control.tsx:
```diff
@@ -306,21 +306,25 @@ class OpenCVControlComponent extends React.PureComponent<Props & DispatchToProps
     private disableImageModifier(alias: string):void {
         const { activeImageModifiers } = this.state;
         const index = activeImageModifiers.findIndex((imageModifier) => imageModifier.alias === alias);
+        const canvas: HTMLCanvasElement | undefined = window.document.getElementById('cvat_canvas_background');
```
I am not sure that writing this code in this method is the correct way.
disableImageModifier can get different modifiers in general, not only equalization. Some can be disabled, some can be enabled. The same about enableImageModifier. 

I suppose it would be better to write this code in ``components/annotation-page/standard-workspace/controls-side-bar/opencv-control.tsx::renderImageContent``.

But on the other hand, adding this CSS class does not guarantee us that the equalization code works (in both places, actually). With the same level of success, we can track the class list of the equalization button. Probably better to use screenshot testing here, because we know that the gray image will be changed significantly after equaliation.

@klakhov, @dvkruchinin What do you think?

bsekachev replied:
As I said without screenshots comparison, the activated button does not guarantee that the algorithm works.

bsekachev replied:
Another option is to show a notification if something goes wrong. As I understand, we do not have such the notification now. @klakhov could help with adding it.

klakhov replied:
I agree that its not the best way to determine that histogram was equalized. But does screenshots comparison work well enough to do it? I'm not sure but maybe it will fail from time to time.
It seems that notification or class is the safest choice.

bsekachev replied:
Generally, typical screenshot is changed significantly, I do not expect the test to be failed time to time. 

![Screenshot from 2021-08-05 15-49-47](https://user-images.githubusercontent.com/40690378/128352671-880f65fe-ee26-4795-91ff-99fb676e9c42.png)
![Screenshot from 2021-08-05 15-49-55](https://user-images.githubusercontent.com/40690378/128352674-da2d1b0d-34b3-4e41-b1b7-3a87a56577eb.png)

But if you believe it is not the best option, okay. In this case, could you add a notification? @klakhov

klakhov replied:
Yes, I think its better to add something like this

bsekachev commented on tests/cypress/integration/actions_tasks2/case_101_opencv_basic_actions.js:
```diff
@@ -0,0 +1,128 @@
+        { x: 400, y: 550 },
+    ];
+
+    function generateString(countPointsToMove) {
+        let action = '';
+        for (let i = 0; i < countPointsToMove; i++) {
+            action += '{rightarrow}';
+        }
+        return action;
```
This function is already defined three times in ``tests/cypress/integration/**/*.ts`` (this is the 4th).
Maybe put it to ``utils.js`` file?

dvkruchinin replied:
Done. The corresponding tests are adapted for this.

bsekachev commented on tests/cypress/integration/actions_tasks2/case_101_opencv_basic_actions.js:
```diff
@@ -0,0 +1,128 @@
+        });
+
+        it('Create a shape with "Intelligent cissors".', () => {
+            cy.opencvCreateShape(createOpencvShape);
+        });
+
+        it('Create a shape with "Intelligent cissors" and "Done" button.', () => {
+            cy.opencvCreateShape(createOpencvShapeSecondLabel);
```
Let's unite it into one test. In both cases, we test creation, just in different scenarios.

dvkruchinin replied:
Done.

bsekachev commented on tests/cypress/support/commands_opencv.js:
```diff
@@ -0,0 +1,52 @@
+});
+
+Cypress.Commands.add('opncvCheckObjectParameters', (objectType) => {
+    cy.get('.cvat-opencv-control-popover').should('be.hidden');
```
Let's remove this check. It might be changed in the future.

dvkruchinin replied:
Removed.

bsekachev commented on tests/cypress/support/commands_opencv.js:
```diff
@@ -0,0 +1,52 @@
+            listCanvasShapeId.push(listCanvasShape[i].id.match(/\d+$/));
+        }
+        const maxId = Math.max(...listCanvasShapeId);
+        cy.get(`#cvat_canvas_shape_${maxId}`).should('exist').and('be.visible');
```
It 100% exists, because maxId was received in the code above from ids of ``cvat_canvas_shape``.

dvkruchinin replied:
Done. Left only the visibility check.

klakhov commented:
@dvkruchinin I've added error notification in case there is equalization error. It looks like that. I think you can remove CSS class in enable/disable modifier functions and check that equalization throws no error now.
![image](https://user-images.githubusercontent.com/50956430/128467339-fe6e0ae8-688b-4b1c-a9e7-e821c52af84b.png)

dvkruchinin commented:
> I've added error notification in case there is equalization error. It looks like that. I think you can remove CSS class in enable/disable modifier functions and check that equalization throws no error now.

Great! I added a class to the notification and reworked the test to check for its absence.
